### PR TITLE
Basic File Storage Impl

### PR DESCRIPTION
**Pull request by evanxg852000**
### Context / purpose
The quickwit-cli needs to work on both s3 and file system storage. this includes a File Storage implementation in the storage crate. https://github.com/quickwit-inc/quickwit/issues/20

### Description
Added an implementation of the File Storage 

### How was this PR tested?
Unit test have been written, thought not much to be done since the abstraction was already well in place. 

### Checklist
- [x] tested locally
- [x] added unit tests
- [x] included documentation
- [x] proof read

### CODE REVIEW

File: storage/src/file_storage.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    fn uri(&self) -> String {
+        format!("file://{}", self.root.to_string_lossy())
+    }
+}
+
+/// A File storage resolver
+#[derive(Debug, Clone)]
+pub struct FileStorageFactory {
+    protocol: String,
```
Comment (fulmicoton) [inline storage/src/file_storage.rs:120]: we don't need a protocol member.

File: storage/src/file_storage.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    fn default() -> Self {
+        FileStorageFactory {
+            protocol: "file".to_string(),
+        }
+    }
+}
+
+impl StorageFactory for FileStorageFactory {
+    fn protocol(&self) -> String {
+        self.protocol.clone()
```
Comment (fulmicoton) [inline storage/src/file_storage.rs:133]: "file".to_string()

File: storage/src/file_storage.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+impl FileStorage {
+    /// Creates a file storage instance given a root path.
+    pub fn new(root: PathBuf) -> Self {
+        Self { root }
+    }
+
+    /// Creates a file storage instance given a uri
+    pub fn from_uri(uri: &str) -> StorageResult<FileStorage> {
```
Comment (fulmicoton) [inline storage/src/file_storage.rs:52]: Im ok with that...
I think I would have prefered something that takes a filepath however, and extract the filepath from the uri in the Factory instead.

Comment (fulmicoton) [reply storage/src/file_storage.rs:52]: Ah you also have the FileStorage::new, so no need to change anything I think.

File: storage/src/file_storage.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            .unwrap();
+        assert_eq!(err.kind(), StorageErrorKind::DoesNotExist);
+
+        let err = file_storage_factory.resolve("test://").err().unwrap();
+        assert_eq!(err.kind(), StorageErrorKind::DoesNotExist);
+        Ok(())
+    }
+
+    // #[tokio::test]
+    // async fn test_file_storage_builder() -> anyhow::Result<()> {
```
Comment (fulmicoton) [inline storage/src/file_storage.rs]: no need for a builder.

File: storage/src/storage_resolver.rs [hunk 5]
```diff
@@ -48,6 +48,7 @@ impl Default for StorageUriResolver {
             per_protocol_resolver: Default::default(),
         };
         resolver.register(RamStorageFactory::default());
+        resolver.register(FileStorageFactory::default());
```
Comment (fulmicoton) [inline storage/src/storage_resolver.rs:51]: sweet you even thought about adding the FileStorageFactory !

File: storage/src/file_storage.rs (hunk 4 above)
Comment (evanxg852000) [reply storage/src/file_storage.rs]: I was still cleaning it. so I just pushed my latest proofread

**Review by guilload — DISMISSED** [review]

**Comment by guilload** [comment]
LGTM.

I suggest we improve the naming with something like `LocalFileStorage,` which is more explicit IMO. I think we also return `StorageErrorKind::DoesNotExist` inappropriately in some cases. We probably need more error types (`InvalidInput`?). That's not something specific to your PR, though. We can fix that later.

**Comment by evanxg852000** [comment]
Thanks @guilload for the feedback. I will discuss this @fulmicoton.

File: storage/src/local_file_storage.rs [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        let full_path = self.root.join(path);
+        fs::copy(full_path, output_path).await?;
+        Ok(())
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<Vec<u8>> {
+        let full_path = self.root.join(path);
+        let mut file = fs::File::open(full_path).await.map_err(|_| {
+            StorageErrorKind::DoesNotExist.with_error(anyhow::anyhow!(
```
Comment (fulmicoton) [inline storage/src/local_file_storage.rs:89]: can you implement a From<io::Error> for storage error that has specific logic for the DoesNotExist case based on the io::ErrorKind.

File: storage/src/error.rs [hunk 7]
```diff
@@ -111,9 +111,9 @@ impl StorageError {
 
 impl From<io::Error> for StorageError {
     fn from(err: io::Error) -> StorageError {
-        StorageError {
-            kind: StorageErrorKind::Io,
-            source: anyhow::Error::from(err),
+        match err.kind() {
+            io::ErrorKind::NotFound => StorageErrorKind::DoesNotExist.with_error(err),
```
Comment (fulmicoton) [inline storage/src/error.rs:115]: yes

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**